File: apps/fsharp/tests/MachineLearning.Tests/Chapter09/BikeWeatherTest.fs
module MachineLearning.Tests.Chapter09.BikeWeatherTest

open System.IO
open System.Text
open Xunit
open MachineLearning.Chapter09.BikeWeather

[<Fact>]
let ``Shift_JIS の CSV を読み込み、天気の番号から名前への Map にする`` () =
    let csvFile = Path.GetTempFileName()
    File.WriteAllText(csvFile, "weather_id,weather\n1,晴れ\n2,曇り\n", shiftJis ())

    Assert.Equal<Map<int, string>>(Map.ofList [ 1, "晴れ"; 2, "曇り" ], loadWeather csvFile)

[<Fact>]
let ``学習用テスト: Shift_JIS のファイルを UTF-8 として読むと文字化けする`` () =
    let csvFile = Path.GetTempFileName()
    File.WriteAllText(csvFile, "晴れ", shiftJis ())

    Assert.NotEqual<string>("晴れ", File.ReadAllText(csvFile, Encoding.UTF8))

let bike =
    [
        {
            Day = "2011-01-01"
            WeatherId = 1
            Count = 100
        }
        {
            Day = "2011-01-02"
            WeatherId = 2
            Count = 50
        }
        {
            Day = "2011-01-03"
            WeatherId = 1
            Count = 300
        }
        {
            Day = "2011-01-04"
            WeatherId = 9
            Count = 10
        }
    ]

[<Fact>]
let ``天気の番号で天気の名前を結合し、名前が無い行は除く`` () =
    let joined = joinWeather (Map.ofList [ 1, "晴れ"; 2, "曇り" ]) bike

    Assert.Equal<(string * int) list>([ "晴れ", 100; "曇り", 50; "晴れ", 300 ], joined |> List.map (fun (w, b) -> w, b.Count))

[<Fact>]
let ``天気ごとの平均利用者数を、多い順に並べる`` () =
    let joined = joinWeather (Map.ofList [ 1, "晴れ"; 2, "曇り" ]) bike

    Assert.Equal<(string * float) list>([ "晴れ", 200.0; "曇り", 50.0 ], meanCountByWeather joined)

File: apps/java/src/test/java/chapter09/BikeWeatherTest.java
package chapter09;

import static org.assertj.core.api.Assertions.assertThat;

import chapter02.Row;
import chapter02.Table;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

class BikeWeatherTest {
  @TempDir Path directory;

  private static Table table(List<String> columns, List<Map<String, String>> rows) {
    return new Table(columns, rows.stream().map(Row::new).toList());
  }

  @Test
  @DisplayName("タブ区切りのファイルを読み込む")
  void loadsTsv() throws IOException {
    Path tsvFile = directory.resolve("bike.tsv");
    Files.writeString(tsvFile, "dteday\tweather_id\tcnt\n2030-04-01\t1\t120\n");

    Table bike = BikeWeather.loadBike(tsvFile);

    assertThat(bike.columns()).containsExactly("dteday", "weather_id", "cnt");
    assertThat(bike.rows().getFirst().text("weather_id")).isEqualTo("1");
    assertThat(bike.rows().getFirst().text("cnt")).isEqualTo("120");
  }

  @Test
  @DisplayName("Shift_JIS のファイルを読み込む")
  void loadsShiftJis() throws IOException {
    Path csvFile = directory.resolve("weather.csv");
    Files.writeString(csvFile, "weather_id,weather\n1,晴れ\n", Charset.forName("Shift_JIS"));

    Table weather = BikeWeather.loadWeather(csvFile);

    assertThat(weather.rows().getFirst().text("weather")).isEqualTo("晴れ");
  }

  @Test
  @DisplayName("Shift_JIS のファイルを UTF-8 として読むと例外になる")
  void shiftJisIsNotUtf8() throws IOException {
    Path csvFile = directory.resolve("weather.csv");
    Files.writeString(csvFile, "weather_id,weather\n1,晴れ\n", Charset.forName("Shift_JIS"));

    org.assertj.core.api.Assertions.assertThatThrownBy(
            () -> DelimitedFiles.load(csvFile, StandardCharsets.UTF_8, ","))
        .isInstanceOf(java.nio.charset.MalformedInputException.class);
  }

  @Test
  @DisplayName("天気 ID で天気の名前を結合する")
  void joinsWeather() {
    Table bike =
        table(
            List.of("weather_id", "cnt"),
            List.of(
                Map.of("weather_id", "2", "cnt", "80"), Map.of("weather_id", "1", "cnt", "120")));
    Table weather =
        table(
            List.of("weather_id", "weather"),
            List.of(
                Map.of("weather_id", "1", "weather", "晴れ"),
                Map.of("weather_id", "2", "weather", "曇り")));

    Table joined = BikeWeather.joinWeather(bike, weather);

    assertThat(joined.columns()).containsExactly("weather_id", "cnt", "weather");
    assertThat(joined.rows()).extracting(r -> r.text("cnt")).containsExactly("80", "120");
    assertThat(joined.rows()).extracting(r -> r.text("weather")).containsExactly("曇り", "晴れ");
  }

  @Test
  @DisplayName("天気の表に無い天気 ID の行は残さない")
  void dropsUnknownWeather() {
    Table bike =
        table(
            List.of("weather_id", "cnt"),
            List.of(
                Map.of("weather_id", "1", "cnt", "120"), Map.of("weather_id", "9", "cnt", "30")));
    Table weather =
        table(
            List.of("weather_id", "weather"), List.of(Map.of("weather_id", "1", "weather", "晴れ")));

    Table joined = BikeWeather.joinWeather(bike, weather);

    assertThat(joined.rows()).extracting(r -> r.text("cnt")).containsExactly("120");
  }

  @Test
  @DisplayName("天気ごとの平均利用者数を多い順に求める")
  void meanCountByWeather() {
    Table joined =
        table(
            List.of("weather", "cnt"),
            List.of(
                Map.of("weather", "雨", "cnt", "20"),
                Map.of("weather", "晴れ", "cnt", "100"),
                Map.of("weather", "晴れ", "cnt", "140")));

    assertThat(BikeWeather.meanCountByWeather(joined))
        .containsExactly(Map.entry("晴れ", 120.0), Map.entry("雨", 20.0));
  }
}
